File: Data Structures & Algorithms/is-palindrome/submission-9.cs
public class Solution {
    // I will a two pointer where the left pointer pointing at first element in the array
    // and the right pointer pointing at the last element. 
    // As I iterate through the characters, I will check if any of the values at the pointers 
    // are non-alphanumeric and skip them if they by moving the pointers.
    // If any of the values at the pointers are not equal, I will return false.
    // If they are equal, I will move the pointers inward.

    // Time Complexity is O(n) where n is the number of chars in the string
    // Space Complexity is O(1) since the algorithm does not require any extra memory

    public bool IsPalindrome(string s) {
        var l = 0;
        var r = s.Length - 1;

        while (l < r) {
            while (l < r && !char.IsLetterOrDigit(s[l])) l++;
            while (l < r && !char.IsLetterOrDigit(s[r])) r--;

            Console.WriteLine(s[l]);
            Console.WriteLine(s[r]);

            if (char.ToLower(s[l]) != char.ToLower(s[r])) return false;

            l++;
            r--;
        }

        return true;
    }
}

File: Data Structures & Algorithms/two-integer-sum-ii/submission-9.js
class Solution {
    // I will use a two pointer algorithm where the left pointer start at the first element of the 
    // array and the right pointer will start at the last element. 
    // As I scan through the array, I will check if the sum of the values at the pointers is less than 
    // the target, inwhich case I will move the left pointer to the left. If it is greater than the 
    // target, I will move the right pointer to the left. Until I find the indices and return them.
    
    // Time Complexity is linear or O(n)
    // Space Complexity is constant or O(1)

    /**
     * @param {number[]} numbers
     * @param {number} target
     * @return {number[]}
     */
    twoSum(numbers, target) {
        let l = 0;
        let r = numbers.length - 1;

        while (l < r) {
            const sum = numbers[l] + numbers[r];
            
            if (sum === target) return [l + 1, r + 1];

            if (sum < target) l++;
            if (sum > target) r--;
        }        

        return [];
    }
}
